content/en/docs/reference/access-authn-authz/validating-admission-policy.md

```diff
@@ -365,3 +366,122 @@ HTTP response code, are used in the HTTP response to the client.
 The currently supported reasons are: `Unauthorized`, `Forbidden`, `Invalid`, `RequestEntityTooLarge`.
 If not set, `StatusReasonInvalid` is used in the response to the client.
 
+### Message Expression
+
+To return a more friendly message when the policy rejects a request, we can use a CEL expression
+to composite a message with `spec.validations[i].messageExpression`. Similar to the validation expression,
+a message expression has access to `object`, `oldObject`, `request`, and `params`. Unlike validations,
+message expression must evaluate to a string.
+
+For example, to better inform the user of the reason of denial when the policy refers to a param, we can have the following validation:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "deploy-replica-policy.example.com"
+spec:
+  paramKind:
+    group: rules.example.com
+    kind: ReplicaLimit
+    version: v1
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+  validations:
+  - expression: "object.spec.replicas <= params.maxReplicas"
+    messageExpression: "'object.spec.replicas must be no greater than ' + string(params.maxReplicas)"
+    reason: Invalid
+```
+
+After creating a params object that limits the replicas to 3 and setting up the binding, when we try to create a deployment with 5 replicas,
+we will receive the following message.
+```
+$ kubectl create deploy --image=nginx nginx --replicas=5
+error: failed to create deployment: deployments.apps "nginx" is forbidden: ValidatingAdmissionPolicy 'deploy-replica-policy.example.com' with binding 'demo-binding-test.example.com' denied request: object.spec.replicas must be no greater than 3
+```
+This is more informative than a static message of "too many replicas".
+
+The message expression takes precedence of the static message defined in `spec.validations[i].message` if both are defined.
+However, in the situation where the message expression fails to evaluate, the result will fall back to use the static message instead.
+
+### Type Checking
+A common error when authoring a policy is type confusion. When a policy definition is being created or updated,
```

Comment:
To establish context, it might also be worth noting here that the CEL expressions are parsed when a ValidatingAdmissionPolicy is created/updated and any gramatical errors are reported at that time, but the fields and types of fields on object, oldObject and params are not checked for correctness..

Reply:
(added)